### PR TITLE
[BugFix] Fix bug specify partition query does not work for list partition

### LINKED ISSUE #17940: [list partition] not worked when select with partition
### Steps to reproduce the behavior (Required)
````sql
CREATE TABLE t1 (
                          dt date not null,
                          num int,
                          city varchar(20),
                          name varchar(20)
                        ) ENGINE=OLAP
                        DUPLICATE KEY(dt)
                        COMMENT "OLAP"
                        PARTITION BY LIST (dt) (
                           PARTITION p1 VALUES IN (("2022-04-01")),
                           PARTITION p2 VALUES IN (("2022-04-02")),
                           PARTITION p3 VALUES IN (("2022-04-03"))
                        )
                        DISTRIBUTED BY HASH(dt) BUCKETS 3
                        PROPERTIES (
                        "replication_num" = "1"
                        );
INSERT INTO t1 PARTITION(p2) VALUES ('2022-04-02', 1, null, 'jack');
SELECT * FROM t1 PARTITION(p1);
````

### Expected behavior (Required)
`Empty set (0.01 sec)`
### Real behavior (Required)
```
+------------+------+------+------+
| dt         | num  | city | name |
+------------+------+------+------+
| 2022-04-02 |    1 | NULL | jack |
+------------+------+------+------+
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
`MAIN-RELEASE-f256177`

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #17940

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java [hunk 1]
```diff
@@ -126,11 +128,14 @@ public List<Long> prune() throws AnalysisException {
...
         Preconditions.checkArgument(columnToPartitionValuesMap.size() == columnToNullPartitions.size());
         Preconditions.checkNotNull(partitionConjuncts);
-        if (columnToPartitionValuesMap.isEmpty() && columnToNullPartitions.isEmpty()) {
+        if (columnToPartitionValuesMap.isEmpty()) {
             // no partition columns, notEvalConjuncts is same with conjuncts
             noEvalConjuncts.addAll(partitionConjuncts);
             return null;
         }
+        if (specifyPartitionIds != null && !specifyPartitionIds.isEmpty()) {
+            return specifyPartitionIds;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java:137]: Returns the intersection of specifyPartitionIds and matches

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 11 / 13 (84.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java|8|10|80.00%|[168, 175]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java|2|2|100.00%|[]|

**Review by wyb — APPROVED** [review]

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**